File: jambeth/jambeth-ioc/src/main/java/de/osthus/ambeth/typeinfo/MethodPropertyInfo.java
package de.osthus.ambeth.typeinfo;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.lang.reflect.Type;

import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.objectcollector.IThreadLocalObjectCollector;
import de.osthus.ambeth.util.ParamChecker;
import de.osthus.ambeth.util.ReflectUtil;

public class MethodPropertyInfo extends AbstractPropertyInfo
{
	protected static final Object[] EMPTY_ARGS = new Object[0];

	protected Method getter, setter;

	protected boolean writable, readable;

	public MethodPropertyInfo(Class<?> entityType, String propertyName, Method getter, Method setter)
	{
		this(entityType, propertyName, getter, setter, null);
	}

	public MethodPropertyInfo(Class<?> entityType, String propertyName, Method getter, Method setter, IThreadLocalObjectCollector objectCollector)
	{
		super(entityType, objectCollector);
		name = propertyName;
		if (name.isEmpty())
		{
			throw new RuntimeException("Not a property method: " + entityType.getName() + "." + getter.getName());
		}
		if (getter != null)
		{
			ParamChecker.assertTrue(!void.class.equals(getter.getReturnType()), "getter");
			ParamChecker.assertTrue(getter.getParameterTypes().length == 0, "getter");
		}
		if (setter != null)
		{
			ParamChecker.assertTrue(setter.getParameterTypes().length == 1, "setter");
		}
		this.getter = getter;
		this.setter = setter;
		writable = this.setter != null && (Modifier.isPublic(this.setter.getModifiers()) || Modifier.isProtected(this.setter.getModifiers()));
		readable = this.getter != null && (Modifier.isPublic(this.getter.getModifiers()) || Modifier.isProtected(this.getter.getModifiers()));
		init(objectCollector);
	}

	@Override
	protected void init(IThreadLocalObjectCollector objectCollector)
	{
		if (entityType == null)
		{
			throw new IllegalArgumentException("No class given");
		}
		if (getter == null && setter == null)
		{
			throw new IllegalArgumentException("No property methods (class is '" + entityType + "')");
		}
		else if (getter != null)
		{
			Class<?> declaringClass = getter.getDeclaringClass();
			propertyType = getter.getReturnType();
			Type returnType = getter.getGenericReturnType();
			elementType = TypeInfoItemUtil.getElementTypeUsingReflection(propertyType, returnType);

			String nameLower = name.toLowerCase();
			Field[] fields = ReflectUtil.getDeclaredFieldsInHierarchy(declaringClass);
			Field backingField = null, weakBackingField = null;
			for (int a = fields.length; a-- > 0;)
			{
				Field field = fields[a];
				int fieldModifiers = field.getModifiers();
				if (Modifier.isStatic(fieldModifiers))
				{
					continue;
				}
				String fieldName = field.getName().toLowerCase();
				if (fieldName.equals(nameLower))
				{
					backingField = field;
					break;
				}
				else if (fieldName.endsWith(nameLower))
				{
					if (weakBackingField != null)
					{
						weakBackingField = null;
						break;
					}
					weakBackingField = field;
				}
			}
			if (backingField == null)
			{
				backingField = weakBackingField;
			}
			this.backingField = backingField;
			if (backingField != null)
			{
				modifiers = backingField.getModifiers();
				putAnnotations(backingField);
			}
			putAnnotations(getter);
			if (setter != null)
			{
				if (setter.getParameterTypes().length != 1 || !setter.getParameterTypes()[0].equals(propertyType))
				{
					throw new RuntimeException("Misfitting property methods for property '" + name + "' on class '" + entityType.getName() + "'");
				}
				putAnnotations(setter);
			}
		}
		else
		{
			Class<?> declaringClass = setter.getDeclaringClass();
			Field[] fields = ReflectUtil.getDeclaredFieldsInHierarchy(declaringClass);
			for (int a = fields.length; a-- > 0;)
			{
				Field field = fields[a];
				int fieldModifiers = field.getModifiers();
				if (Modifier.isFinal(fieldModifiers) || Modifier.isStatic(fieldModifiers))
				{
					continue;
				}
				String fieldName = field.getName();
				if (fieldName.endsWith(name))
				{
					if (backingField != null)
					{
						backingField = null;
						break;
					}
					backingField = field;
				}
			}
			if (backingField != null)
			{
				modifiers = backingField.getModifiers();
			}
			propertyType = setter.getParameterTypes()[0];
			Type[] paramTypes = setter.getGenericParameterTypes();
			elementType = TypeInfoItemUtil.getElementTypeUsingReflection(propertyType, paramTypes[0]);
			putAnnotations(setter);
		}
		if (getter != null && Modifier.isNative(getter.getModifiers()))
		{
			modifiers |= Modifier.NATIVE;
		}
		if (setter != null && Modifier.isNative(setter.getModifiers()))
		{
			modifiers |= Modifier.NATIVE;
		}
		if (setter != null)
		{
			modifiers &= ~Modifier.FINAL;
		}
		Boolean getterIsPublic = getter != null ? Boolean.valueOf(Modifier.isPublic(getter.getModifiers())) : null;
		Boolean setterIsPublic = setter != null ? Boolean.valueOf(Modifier.isPublic(setter.getModifiers())) : null;
		if (Boolean.TRUE.equals(getterIsPublic) || Boolean.TRUE.equals(setterIsPublic))
		{
			modifiers &= ~Modifier.PRIVATE;
			modifiers &= ~Modifier.PROTECTED;
			modifiers |= Modifier.PUBLIC;
		}
		else
		{
			Boolean getterIsProtected = getter != null ? Boolean.valueOf(Modifier.isProtected(getter.getModifiers())) : null;
			Boolean setterIsProtected = setter != null ? Boolean.valueOf(Modifier.isProtected(setter.getModifiers())) : null;
			if (Boolean.TRUE.equals(getterIsProtected) || Boolean.TRUE.equals(setterIsProtected))
			{
				modifiers &= ~Modifier.PRIVATE;
				modifiers &= ~Modifier.PUBLIC;
				modifiers |= Modifier.PROTECTED;
			}
			else
			{
				modifiers &= ~Modifier.PROTECTED;
				modifiers &= ~Modifier.PUBLIC;
				modifiers |= Modifier.PRIVATE;
			}
		}

		refreshDeclaringType();
		super.init(objectCollector);
	}

	protected void refreshDeclaringType()
	{
		Class<?> fieldDC = backingField != null ? backingField.getDeclaringClass() : null;
		Class<?> getterDC = getter != null ? getter.getDeclaringClass() : null;
		Class<?> setterDC = setter != null ? setter.getDeclaringClass() : null;
		Class<?> mostSpecificDC = fieldDC;
		if (mostSpecificDC == null)
		{
			mostSpecificDC = getterDC;
		}
		else if (getterDC != null)
		{
			mostSpecificDC = getterDC.isAssignableFrom(mostSpecificDC) ? mostSpecificDC : getterDC;
		}
		if (mostSpecificDC == null)
		{
			mostSpecificDC = setterDC;
		}
		else if (setterDC != null)
		{
			mostSpecificDC = setterDC.isAssignableFrom(mostSpecificDC) ? mostSpecificDC : setterDC;
		}
		declaringType = mostSpecificDC;
	}

	@Override
	public void refreshAccessors(Class<?> realType)
	{
		super.refreshAccessors(realType);
		getter = ReflectUtil.getDeclaredMethod(true, realType, getPropertyType(), "get" + getName());
		if (getter == null)
		{
			getter = ReflectUtil.getDeclaredMethod(true, realType, getPropertyType(), "is" + getName());
		}
		setter = ReflectUtil.getDeclaredMethod(true, realType, null, "set" + getName(), getPropertyType());
		writable = setter != null && (Modifier.isPublic(setter.getModifiers()) || Modifier.isProtected(setter.getModifiers()));
		readable = getter != null && (Modifier.isPublic(getter.getModifiers()) || Modifier.isProtected(getter.getModifiers()));
		refreshDeclaringType();
	}

	@Override
	public boolean isWritable()
	{
		return writable;
	}

	@Override
	public boolean isReadable()
	{
		return readable;
	}

	public Method getGetter()
	{
		return getter;
	}

	public Method getSetter()
	{
		return setter;
	}

	@Override
	public Object getValue(Object obj)
	{
		if (getter == null)
		{
			return null;
		}
		try
		{
			return getter.invoke(obj, EMPTY_ARGS);
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e, "Error occured while calling '" + getter + "' on object '" + obj + "' of type '" + obj.getClass().toString()
					+ "'");
		}
	}

	@Override
	public void setValue(Object obj, Object value)
	{
		if (setter == null)
		{
			throw new UnsupportedOperationException("No setter configured for property " + name);
		}
		Object[] args = { value };
		try
		{
			setter.invoke(obj, args);
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e, "Error occured while calling '" + setter + "' on object '" + obj + "' of type '" + obj.getClass().toString()
					+ "' with argument '" + value + "'");
		}
	}
}

File: ambeth/Ambeth.Util/ambeth/typeinfo/MethodPropertyInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using De.Osthus.Ambeth.Util;
using System.Reflection;
using De.Osthus.Ambeth.Exceptions;

namespace De.Osthus.Ambeth.Typeinfo
{
    public class MethodPropertyInfo : AbstractPropertyInfo
    {
        protected static readonly Object[] EMPTY_ARGS = new Object[0];

        public MethodInfo Getter { get; protected set; }

        public MethodInfo Setter { get; protected set; }

        public MethodPropertyInfo(Type entityType, String propertyName, MethodInfo getter, MethodInfo setter)
            : base(entityType)
        {
            if (propertyName.Length == 0)
            {
                throw new Exception("Not a property method: " + entityType.Name + "." + getter.Name);
            }

            EntityType = entityType;
            Name = propertyName;

            if (getter != null)
            {
                ParamChecker.AssertTrue(!typeof(void).Equals(getter.ReturnType), "getter");
                ParamChecker.AssertTrue(getter.GetParameters().Length == 0, "getter");
            }
            if (setter != null)
            {
                ParamChecker.AssertTrue(setter.GetParameters().Length == 1, "setter");
            }
            Getter = getter;
            Setter = setter;
            IsWritable = setter != null && setter.IsPublic;
            IsReadable = getter != null && getter.IsPublic;

            Init();
        }

        protected void RefreshDeclaringType()
	    {
		    Type fieldDC = BackingField != null ? BackingField.DeclaringType : null;
            Type getterDC = Getter != null ? Getter.DeclaringType : null;
            Type setterDC = Setter != null ? Setter.DeclaringType : null;
            Type mostSpecificDC = fieldDC;
		    if (mostSpecificDC == null)
		    {
			    mostSpecificDC = getterDC;
		    }
		    else if (getterDC != null)
		    {
			    mostSpecificDC = getterDC.IsAssignableFrom(mostSpecificDC) ? mostSpecificDC : getterDC;
		    }
		    if (mostSpecificDC == null)
		    {
			    mostSpecificDC = setterDC;
		    }
		    else if (setterDC != null)
		    {
                mostSpecificDC = setterDC.IsAssignableFrom(mostSpecificDC) ? mostSpecificDC : setterDC;
		    }
		    DeclaringType = mostSpecificDC;
	    }

        protected override void Init()
        {
            if (EntityType == null)
            {
                throw new ArgumentException("No class given");
            }
            if (Getter == null && Setter == null)
            {
                throw new ArgumentException("No property methods (class is '" + EntityType + "')");
            }
            else if (Getter != null)
            {
                Type declaringClass = Getter.DeclaringType;
                PropertyType = Getter.ReturnType;
                ElementType = TypeInfoItemUtil.GetElementTypeUsingReflection(PropertyType, null);

                String nameLower = Name.ToLowerInvariant();
                FieldInfo[] fields = ReflectUtil.GetDeclaredFieldsInHierarchy(declaringClass);
                FieldInfo backingField = null;
                FieldInfo weakBackingField = null;
                String backingFieldName = "<" + Name + ">k__BackingField";

                for (int a = fields.Length; a-- > 0; )
                {
                    FieldInfo field = fields[a];
                    if (field.IsStatic)
                    {
                        continue;
                    }
                    String fieldName = field.Name.ToLowerInvariant();
                    if (fieldName.Equals(nameLower) || field.Name.Equals(backingFieldName))
                    {
                        backingField = field;
                        break;
                    }
                    else if (fieldName.EndsWith(nameLower))
                    {
                        if (weakBackingField != null)
                        {
                            weakBackingField = null;
                            break;
                        }
                        weakBackingField = field;
                    }
                }
                if (backingField == null)
                {
                    backingField = weakBackingField;
                }
                BackingField = backingField;
                if (backingField != null)
                {
                    //PutAnnotations(backingField);
                }
                PutAnnotations(Getter);
                if (Setter != null)
                {
                    if (Setter.GetParameters().Length != 1 || !PropertyType.IsAssignableFrom(Setter.GetParameters()[0].ParameterType))
                    {
                        throw new Exception("Misfitting property methods for property '" + Name + "' on class '" + EntityType.Name + "'");
                    }
                    PutAnnotations(Setter);
                }
            }
            else
            {
                Type declaringClass = Setter.DeclaringType;
                FieldInfo[] fields = ReflectUtil.GetDeclaredFieldsInHierarchy(declaringClass);
                for (int a = fields.Length; a-- > 0; )
                {
                    FieldInfo field = fields[a];
                    if (field.IsInitOnly || field.IsStatic)
                    {
                        continue;
                    }
                    String fieldName = field.Name;
                    if (fieldName.EndsWith(Name))
                    {
                        if (BackingField != null)
                        {
                            BackingField = null;
                            break;
                        }
                        BackingField = field;
                    }
                }
                PropertyType = Setter.GetParameters()[0].ParameterType;
                ElementType = TypeInfoItemUtil.GetElementTypeUsingReflection(PropertyType, null);
                PutAnnotations(Setter);
            }
            RefreshDeclaringType();
            base.Init();
        }

        public override void RefreshAccessors(Type realType)
	    {
            base.RefreshAccessors(realType);
		    Getter = ReflectUtil.GetDeclaredMethod(true, realType, PropertyType, "get_" + Name);
		    Setter = ReflectUtil.GetDeclaredMethod(true, realType, null, "set_" + Name, PropertyType);
            IsWritable = this.Setter != null && !Setter.IsPrivate;
		    IsReadable = this.Getter != null && !Getter.IsPrivate;
            RefreshDeclaringType();
	    }

        public override Object GetValue(Object obj)
        {
            if (Getter == null)
            {
                return null;
            }
            try
            {
                return Getter.Invoke(obj, EMPTY_ARGS);
            }
            catch (Exception e)
            {
                throw RuntimeExceptionUtil.Mask(e, "Error occured while calling '" + Getter + "' on object '" + obj + "' of type '" + obj.GetType().ToString()
                        + "'");
            }
        }

        public override void SetValue(Object obj, Object value)
        {
            if (Setter == null)
            {
                throw new NotSupportedException("No setter configure for property " + Name);
            }
            Object[] args = { value };
            try
            {
                Setter.Invoke(obj, args);
            }
            catch (Exception e)
            {
                throw RuntimeExceptionUtil.Mask(e, "Error occured while calling '" + Setter + "' on object '" + obj + "' of type '" + obj.GetType().ToString()
                        + "' with argument '" + value + "'");
            }
        }
    }
}